Objects/setobject.c

```diff
@@ -1063,36 +1063,67 @@ make_new_set_basetype(PyTypeObject *type, PyObject *iterable)
 /* The empty frozenset is a singleton */
 static PyObject *emptyfrozenset = NULL;
 
+static PyObject *
+make_new_frozenset(PyTypeObject *type, PyObject *iterable)
+{
+    assert(PyType_Check(type));
+    if (iterable != NULL) {
+        if (PyFrozenSet_CheckExact(iterable)) {
+            /* frozenset(f) is idempotent */
+            Py_INCREF(iterable);
+            return iterable;
+        }
+        PyObject *res = make_new_set((PyTypeObject *)type, iterable);
+        if (res == NULL || PySet_GET_SIZE(res) != 0) {
+            return res;
+        }
+        /* If the created frozenset is empty, return the empty frozenset singleton instead */
+        Py_DECREF(res);
+    }
+
+    // The empty frozenset is a singleton
+    if (emptyfrozenset == NULL) {
+        emptyfrozenset = make_new_set((PyTypeObject *)type, NULL);
+    }
+    Py_XINCREF(emptyfrozenset);
+    return emptyfrozenset;
+}
+
 static PyObject *
 frozenset_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 {
-    PyObject *iterable = NULL, *result;
+    PyObject *iterable = NULL;
 
-    if (type == &PyFrozenSet_Type && !_PyArg_NoKeywords("frozenset", kwds))
+    if (type == &PyFrozenSet_Type && !_PyArg_NoKeywords("frozenset", kwds)) {
         return NULL;
+    }
 
-    if (!PyArg_UnpackTuple(args, type->tp_name, 0, 1, &iterable))
+    if (!PyArg_UnpackTuple(args, type->tp_name, 0, 1, &iterable)) {
         return NULL;
+    }
 
-    if (type != &PyFrozenSet_Type)
+    if (type != &PyFrozenSet_Type) {
         return make_new_set(type, iterable);
+    }
 
-    if (iterable != NULL) {
-        /* frozenset(f) is idempotent */
-        if (PyFrozenSet_CheckExact(iterable)) {
-            Py_INCREF(iterable);
-            return iterable;
-        }
-        result = make_new_set(type, iterable);
-        if (result == NULL || PySet_GET_SIZE(result))
-            return result;
-        Py_DECREF(result);
+    return make_new_frozenset(type, iterable);
```

Comment:
You can replace the 4 lines with: `return make_new_frozenset(type, iterable);`.

Reply:
Oh nice catch!